pbft/pbft.cpp

```diff
@@ -994,6 +1007,533 @@ pbft::get_high_water_mark()
+    for (int i{0}; i < msg.checkpoint_messages_size(); ++i)
+    {
+        const auto checkpoint_msg = msg.checkpoint_messages(i);
+        bzn_envelope original_checkpoint;
+        if (original_checkpoint.ParseFromString(checkpoint_msg)
+            || !this->crypto->verify(original_checkpoint) // TODO: consider removing this - check with debugger
+            || !message.ParseFromString(original_checkpoint.pbft()))
+        {
+            continue;
+        }
+        this->handle_checkpoint(message, original_checkpoint);
+    }
+}
+
+void
+pbft::replica_broadcasts_viewchange(const pbft_msg& msg)
+{
+    // replica sends VIEWCHANGE message
+    pbft_msg viewchange_message{
+            make_viewchange(
+                    msg.view()
+                    , this->latest_stable_checkpoint().first
+                    , this->stable_checkpoint_proof
+                    , this->prepared_operations_since_last_checkpoint())
+    };
+
+    this->view_is_valid = false;
+    this->last_view_sent = msg.view();
+    this->broadcast(this->wrap_message(viewchange_message));
+
+}
+
+void
+pbft::handle_viewchange(const pbft_msg& msg, const bzn_envelope& original_msg)
+{
+    if (this->is_valid_viewchange_message(msg, original_msg))
+    {
+        this->valid_view_change_messages[msg.view()].insert(original_msg.SerializeAsString());
+    }
+    else
+    {
+        LOG(error) << "handle_viewchange - invalid viewchange message, ignoring";
+        return;
+    }
+
+    this->save_checkpoint(msg);
+
+    // When a replica recieves f + 1 view change messages it sends one as well
+    // should it be == or >=  ?
+    if (this->view_is_valid && this->valid_view_change_messages.size() == this->max_faulty_nodes() + 1 && msg.view() > this->last_view_sent)
+    {
+        this->replica_broadcasts_viewchange(msg);
+    }
+
+    // TODO move this to the end...
+    // we want to filter std::map<uint64_t, std::set<bzn_envelope>> valid_view_change_messages for
+    // 1) we would be the primary for that view
+    // 2) that view is greater than the current view
+    // 3) we have 2 f + 1 viewchange messages for it
+    const auto viewchange = std::find_if(this->valid_view_change_messages.begin(), this->valid_view_change_messages.end(), [&](const auto& p)
+            {
+                return ((this->get_primary(p.first).uuid == this->get_uuid()) &&  (p.first > this->view) && (p.second.size() >= 2 * this->max_faulty_nodes() + 1));
+            });
+
+    if (viewchange == this->valid_view_change_messages.end())
+    {
+        LOG (debug) << "handle_viewchange - not enough viewchange messages";
+        return;
+    }
+
+    // When the primary of the new view recieves 2f valid view change messages for the new view
+    // the primary of new view broadcasts NEWVIEW
+
+    // TODO Refactor into method
+    {
+        // create the newview and and broadcast it
+
+        // viewchange->second is a set of bzn_envelope strings
+        // -- we want a of set<pbft_msg>
+
```

Comment:
This is rather confusing. I suggest making pbft::valid_view_change_messages a
std::map<uint64_t, std::map<uuid_t, bzn_envelope>>- so a 2d map keyed by view and then by sender uuid.

Reply:
I've made this change and propagated it through the module.